tao_compiler/mlir/disc/transforms/disc_hlo_legalize_to_lhlo.cc

```diff
@@ -184,6 +186,52 @@ struct HloToLhloCustomCallOpConverter : public BaseOpConversion<CustomCallOp> {
   }
 };
 
+struct HloToLhloCustomCallOpV2Converter
+    : public BaseOpConversion<CustomCallV2Op> {
+ public:
+  using BaseOpConversion<CustomCallV2Op>::BaseOpConversion;
+
+  LogicalResult matchAndRewrite(
+      CustomCallV2Op hloOp, OpAdaptor adaptor,
+      ConversionPatternRewriter& rewriter) const override {
+    Location loc = hloOp->getLoc();
+    SmallVector<Type> resultTypes;
+    for (Value v : hloOp->getResults()) {
+      auto ty = v.getType().cast<RankedTensorType>();
+      resultTypes.push_back(
+          MemRefType::get(ty.getShape(), ty.getElementType()));
+    }
+    auto lhloOp = rewriter.create<lmhlo_disc::CustomCallV2Op>(
+        loc, resultTypes, adaptor.getOperands(), hloOp->getAttrs());
+
+    // Those buffers that are returned by the custom call op should be
+    // deallocated manually. Currently we rely on buffer deallocation pass to
+    // automatically insert dealloc op for each alloc-like op. However the pass
+    // does not support multi-results alloc-like op.
+    //
+    // To solve the above problem, we simply add a copy op for each result
+    // buffer and insert a dealloc op explicitly for the old result buffer when
+    // bufferizing.
+    // TODO(disc): these copy ops may inroduce luanch overhead. Add a pass to
+    // remove these copy ops after the normal buffer deallocaiotn pass.
```

Comment:
looks like a tricky workaround ....

Reply:
It's a community pass and very complicated in order to support more general cases. I prefer to add a pass later to simplify the copy + dealloc later in our scenario.